### PR TITLE
access Ras updated

### PR DESCRIPTION
**Pull request by CJMeek**
Signed-off-by: CJMeek <up887397@myport.ac.uk>

### CODE REVIEW

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java [hunk 1]
```diff
@@ -14,16 +34,108 @@
...
 	protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException {
 		
+		Map<String, String> paramMap = getParameterMap(req);
+		
+		try {
+			
+		List<IRunResult> runs = getRuns(paramMap.get("requestor"), Instant.parse(paramMap.get("to")), 
+				Instant.parse(paramMap.get("from")), paramMap.get("testname"));
+		
+		int pageSize = Integer.parseInt(paramMap.get("pageSize").trim());
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java:52]: what happens if the pagesize is not numeric

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java [hunk 2]
```diff
@@ -14,16 +34,108 @@
...
 	private static final long serialVersionUID = 1L;
 	
 	@Override
 	protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException {
 		
+		Map<String, String> paramMap = getParameterMap(req);
+		
+		try {
+			
+		List<IRunResult> runs = getRuns(paramMap.get("requestor"), Instant.parse(paramMap.get("to")),
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java:49]: What happens if Instant fails to parse the to and from

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java [hunk 3]
```diff
@@ -14,16 +34,108 @@
...
+			JsonObject returnObj = new JsonObject();
+			returnObj.addProperty("pageSize", String.valueOf(pageSize));
+			returnObj.addProperty("pageNumber", String.valueOf(pageNum));
+			returnObj.addProperty("numPages", String.valueOf(numPages));
+			
+			JsonArray returnArray = new JsonArray();
+			
+			for(IRunResult run : newList) {
+				
+				JsonObject runObj = new JsonObject();
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java:74]: Why are you not using a POJO?

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java [hunk 4]
```diff
@@ -14,16 +34,108 @@
...
+		List<IRunResult> runs = getRuns(paramMap.get("requestor"), Instant.parse(paramMap.get("to")), 
+				Instant.parse(paramMap.get("from")), paramMap.get("testname"));
+		
+		int pageSize = Integer.parseInt(paramMap.get("pageSize").trim());
+		
+		int amountOfRuns = runs.size();
+		int startIndex = 0;
+		int endIndex = 0;
+		int pageNum = 1;
+
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java:58]: cant see where you get the page number from the request

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java [hunk 5]
```diff
@@ -14,16 +34,108 @@
...
+		int pageSize = Integer.parseInt(paramMap.get("pageSize").trim());
+		
+		int amountOfRuns = runs.size();
+		int startIndex = 0;
+		int endIndex = 0;
+		int pageNum = 1;
+		
+		int numPages = (int)Math.ceil(amountOfRuns / pageSize);
+		
+		while(pageSize > amountOfRuns) {
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java:61]: think you need to rethink this loop

**Review by Michael-Baylis — CHANGES_REQUESTED** [review]

File: galasa-parent/dev.galasa/src/main/java/dev/galasa/api/run/RunResult.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+	private final String testShortName;
+	private final String bundle;
+	private final String requestor;
+	private final String result;
+	private final String status;
+	private final Instant queued;
+	private final Instant start;
+	private final Instant end;
+	
+	public RunResult(String runName, String testName, String testShortName, String bundle,
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa/src/main/java/dev/galasa/api/run/RunResult.java:19]: You will need the appropriate set* so the class can be used at the other end

File: galasa-parent/dev.galasa.framework.obr/pom.xml [hunk 7]
```diff
@@ -226,6 +226,11 @@
 			<version>0.6.0</version>
 			<scope>compile</scope>
 		</dependency>
+		<dependency>
+    		<groupId>org.apache.commons</groupId>
+    		<artifactId>commons-collections4</artifactId>
+    		<version>4.4</version>
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.obr/pom.xml:232]: scope compile for the obr or won't be included

**Review by Michael-Baylis — APPROVED** [review]

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java [hunk 8]
```diff
@@ -14,16 +42,241 @@
...
+
+			
+		}
+		
+		List<RunResult> runs = new ArrayList<>();
+		
+		try {
+			runs = getRuns(critList);
+		} catch (ResultArchiveStoreException e) {
+			e.printStackTrace();
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java:113]: no, no, no, no.  NEVER use e.printStackTrace.  always throw the exception again

File: galasa-parent/dev.galasa/src/main/java/dev/galasa/api/run/RunResult.java (hunk 6 above)
Comment (CJMeek) [reply galasa-parent/dev.galasa/src/main/java/dev/galasa/api/run/RunResult.java:19]: So should I remove the parameters from the constructor and just have set methods?

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java [hunk 9]
```diff
@@ -14,16 +42,241 @@
...
+				
+				obj.add("runs", tree);
+				
+				returnArray.add(obj);
+				
+				pageIndex+=1;
+			}
+		}
+		
+		if(paramMap.get("pageNum") != null && !paramMap.get("pageNum").equals("")) {
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java:177]: Get all the parameters at the top,   so it is clear where it is coming from

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java [hunk 10]
```diff
@@ -14,16 +42,241 @@
...
+		
+		if(!query.isEmpty()){
+			if(!ExtractQuerySort.isAscending(query, "to")) {
+				runs.sort(Comparator.nullsLast(Comparator.comparing(RunResult::getEnd, Comparator.nullsLast(Comparator.naturalOrder()))).reversed());
+			}
+		}
+		
+		int pageSize = 100;
+		
+		if(paramMap.get("pageSize") != null && !paramMap.get("pageSize").equals("")) {
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java:129]: Get all the parameters at the top of the method, so it is clear where it is coming from,  and what is being defaulted

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunResultUtility.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import dev.galasa.framework.spi.ResultArchiveStoreException;
+
+public class RunResultUtility {
+	
+	
+	public RunResultUtility() {
+
+	}
+	
+	public @NotNull RunResult toRunResult(@NotNull IRunResult runResult) throws ResultArchiveStoreException {
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunResultUtility.java:18]: Static?

**Review by Michael-Baylis — CHANGES_REQUESTED** [review]

File: galasa-parent/dev.galasa/src/main/java/dev/galasa/api/run/RunResult.java (hunk 6 above)
Comment (Michael-Baylis) [reply galasa-parent/dev.galasa/src/main/java/dev/galasa/api/run/RunResult.java:19]: either,  up to you really,  you need the set* for json to work properly.   but you want to maintain the constructor as well.....

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java [hunk 12]
```diff
@@ -14,16 +42,219 @@
...
+				critList.add(toCriteria);
+			}
+			if(from != null) {
+				fromCrit = Instant.parse(from);
+				RasSearchCriteriaQueuedFrom fromCriteria = new RasSearchCriteriaQueuedFrom(fromCrit);
+				critList.add(fromCriteria);
+			}
+		}catch(Exception e) {
+			
+			sendError(resp, "Error parsing Instant", 400);
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java:107]: you don't need to sendError and throw,  the throw will do a send error

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java [hunk 13]
```diff
@@ -14,16 +42,219 @@
...
+		List<IRasSearchCriteria> critList = new ArrayList<>();
+		
+		if(!paramMap.isEmpty()) {
+			
+			if(paramMap.get("pageNum") != null && !paramMap.get("pageNum").equals("")) {
+				try {
+					pageNum = Integer.parseInt(paramMap.get("pageNum"));
+				}catch(Exception e) {
+					
+					sendError(resp, "Error parsing integer", 400);
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java:69]: you don't need to sendError and throw,  the throw will do a send error

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java [hunk 14]
```diff
@@ -14,16 +42,219 @@
...
+			
+		}
+		
+		List<RunResult> runs = new ArrayList<>();
+		
+		try {
+			runs = getRuns(critList);
+		} catch (Exception e) {
+			
+			sendError(resp, "Error retrieving runs", 400);
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java:129]: you don't need to sendError and throw,  the throw will do a send error

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java [hunk 15]
```diff
@@ -14,16 +42,219 @@
...
+		}
+		
+		String json = "";
+		
+		if(returnArray.size() != 0) {
+			try {
+			json = gson.toJson(returnArray.get(pageNum-1));
+			}catch(Exception e) {
+				
+				sendError(resp, "Error retrieving page", 400);
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java:180]: you don't need to sendError and throw,  the throw will do a send error

File: galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java [hunk 16]
```diff
@@ -14,16 +42,219 @@
...
+	try {
+		PrintWriter out = resp.getWriter();
+		resp.setContentType( "Application/json");
+		resp.setHeader("Access-Control-Allow-Origin", "*");
+		out.print(json);
+		out.flush();
 	
+	}catch(Exception e) {
+		
+		sendError(resp, "Error retrieving runs", 400);
```
Comment (Michael-Baylis) [inline galasa-parent/dev.galasa.framework.api.ras/src/main/java/dev/galasa/framework/api/ras/internal/RunQuery.java:196]: you don't need to sendError and throw,  the throw will do a send error

**Review by Michael-Baylis — CHANGES_REQUESTED** [review]
sorry,  didn't think it through.    just throw,  dont send error,  and you can remove the senderror method.  think that is it

**Review by Michael-Baylis — APPROVED** [review]

**Pull request merged**